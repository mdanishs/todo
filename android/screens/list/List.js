import React, { useState } from 'react'
import { ScrollView, StyleSheet } from 'react-native';
import ListItem from './components/ListItem';
import TextControl from './components/TextControl';
import { createTodo } from '../../utils'

const List = () => {
  const [todos, setTodos] = useState([]);

  const handleTaskAdd = (task) => {
    setTodos([...todos, createTodo(task)]);
  }

  const handleStateChanged = (task, checked) => {
    const toEdit = todos.find(t => t.id === task.id);
    if (toEdit) {
      toEdit.completed = checked;
    }
    setTodos([...todos]);
  }

  const handleDelete = (task) => {
    const filtered = todos.filter(t => t.id !== task.id);
    setTodos([...filtered]);
  }

  const renderTodos = () => {
    return todos.map((todo, i) => (
      <ListItem
        key={i}
        text={todo.text}
        checked={todo.completed}
        onStateChange={(checked) => { handleStateChanged(todo, checked) }}
        onDelete={() => { handleDelete(todo) }}
      />
    ))
  }

  return (
    <ScrollView>
      <TextControl
        title="Add"
        placeholder="What's next on your list"
        onSubmit={handleTaskAdd}
      />
      {renderTodos()}
    </ScrollView>
  )
}

export default List;