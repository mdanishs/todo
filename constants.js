export const COLORS = {
  PRIMARY: '#76777B',
  SECONDARY: '#54565B',
  BACKGROUND_LIGHT: '#FAF8EB',
  BACKGROUND_DARK: '#939598',
  TITLE: '#FAD41B',
  WHITE: '#FFFFFF',
}

export const SPACING = {
  MARGIN_5: 5,
  MARGIN_10: 10,
  MARGIN_15: 15,
  MARGIN_20: 20,
  PADDING_5: 5,
  PADDING_10: 10,
  PADDING_15: 15,
  PADDING_20: 20,
}

export const TYPOGRAPHY = {
  TITLE: {
    fontSize: 30,
    fontWeight: 'bold',
  },
  ACTION: {
    fontSize: 17,
    fontWeight: 'bold',
  }
}

export const TEST_IDS = {
  LIST_TEXT_INPUT: 'list-text-input',
  LIST_ADD_BUTTON: 'list-add-button',
  LIST_ITEM_TEXT: 'list-item-text',
  LIST_ITEM_CHECKBOX: 'list-item-checkbox',
  LIST_ITEM_DELETE_BUTTON: 'list-item-delete-button',
}