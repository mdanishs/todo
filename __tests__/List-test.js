import 'react-native';
import React from 'react';
import { fireEvent, render } from '@testing-library/react-native';
import List from '../screens/list/List';
import { TEST_IDS } from '../constants';
let list;

beforeEach(() => {
  list = render(<List />);
})

it('renders list correctly', () => {
  expect(list).toBeTruthy();
});

it('adds and deletes a task to and from list', async () => {
  const { getByTestId, queryByTestId } = list;
  const textInput = getByTestId(TEST_IDS.LIST_TEXT_INPUT);
  const addButton = getByTestId(TEST_IDS.LIST_ADD_BUTTON);

  fireEvent.changeText(textInput, 'My todo 1');
  fireEvent.press(addButton);

  let listItemText = getByTestId(TEST_IDS.LIST_ITEM_TEXT);
  expect(listItemText).toBeDefined();
  expect(listItemText.children[0]).toEqual('My todo 1');

  const deleteButton = getByTestId(TEST_IDS.LIST_ITEM_DELETE_BUTTON);
  fireEvent.press(deleteButton);

  expect(queryByTestId(TEST_IDS.LIST_ITEM_TEXT)).toBeNull();
});

it('marks the task as completed on check', () => {
  const { getByTestId, toJSON } = list;
  const textInput = getByTestId(TEST_IDS.LIST_TEXT_INPUT);
  const addButton = getByTestId(TEST_IDS.LIST_ADD_BUTTON);

  fireEvent.changeText(textInput, 'My todo 1');
  fireEvent.press(addButton);
  
  const checkbox = getByTestId(TEST_IDS.LIST_ITEM_CHECKBOX);
  fireEvent.press(checkbox);

  expect(toJSON()).toMatchSnapshot();
});
