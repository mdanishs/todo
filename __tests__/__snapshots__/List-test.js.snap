// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`marks the task as completed on check 1`] = `
<RCTScrollView>
  <View>
    <View
      style={
        Object {
          "alignItems": "center",
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "padding": 10,
        }
      }
    >
      <TextInput
        allowFontScaling={true}
        onChangeText={[Function]}
        placeholder="What's next on your list"
        rejectResponderTermination={true}
        style={
          Object {
            "backgroundColor": "#FFFFFF",
            "borderRadius": 5,
            "flexGrow": 1,
            "marginRight": 5,
            "paddingLeft": 15,
          }
        }
        testID="list-text-input"
        underlineColorAndroid="transparent"
        value="My todo 1"
      />
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#76777B",
            "borderRadius": 50,
            "display": "flex",
            "height": 50,
            "justifyContent": "center",
            "opacity": 1,
            "width": 50,
          }
        }
        testID="list-add-button"
      >
        <Text
          style={
            Object {
              "color": "#FFFFFF",
              "fontSize": 17,
              "fontWeight": "bold",
            }
          }
        >
          +
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#FAF8EB",
          "borderRadius": 5,
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "margin": 10,
          "marginBottom": 0,
          "padding": 10,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#54565B",
            },
          ]
        }
        testID="list-item-text"
      >
        My todo 1
      </Text>
      <View
        style={
          Object {
            "alignContent": "center",
            "display": "flex",
            "flexDirection": "row",
            "justifyContent": "center",
          }
        }
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "opacity": 1,
            }
          }
          testID="list-item-checkbox"
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#ffc484",
                "borderColor": "#ffc484",
                "borderRadius": 12.5,
                "borderWidth": 1,
                "height": 25,
                "justifyContent": "center",
                "transform": Array [
                  Object {
                    "scale": 0.7,
                  },
                ],
                "width": 25,
              }
            }
          >
            <Image
              source={
                Object {
                  "testUri": "../../react-native-bouncy-checkbox/build/dist/check.png",
                }
              }
              style={
                Array [
                  Object {
                    "height": 10,
                    "width": 10,
                  },
                  undefined,
                ]
              }
            />
          </View>
          <View
            style={
              Object {
                "marginLeft": 16,
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#757575",
                    "fontSize": 16,
                    "textDecorationLine": "line-through",
                  },
                  undefined,
                ]
              }
            />
          </View>
        </View>
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID="list-item-delete-button"
        >
          <Text
            style={
              Object {
                "fontSize": 17,
                "fontWeight": "bold",
                "marginTop": -2,
              }
            }
          >
            x
          </Text>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
